candidate-agreement/src/lib.rs

```
@@ -16,21 +16,610 @@
+		Recovery: MessageRecovery<Context> + 'static,
+		PropagateStatements: Stream<Item=Context::StatementBatch,Error=Err> + 'static,
+		LocalCandidate: IntoFuture<Item=Context::ParachainCandidate> + 'static
+{
+	let (bft_in_in, bft_in_out) = mpsc::unbounded();
+	let (bft_out_in, bft_out_out) = mpsc::unbounded();
+
+	let agreement = {
+		let bft_context = BftContext {
+			context: params.context,
+			table: params.table.clone(),
+			timer: params.timer.clone(),
+			round_timeout_multiplier: params.round_timeout_multiplier,
+		};
+
+		bft::agree(
+			bft_context,
+			params.nodes,
+			params.max_faulty,
+			bft_in_out.map(bft::ContextCommunication).map_err(|_| Error::IoTerminated),
+			bft_out_in.sink_map_err(|_| Error::IoTerminated),
+		)
+	};
+
+	let route_messages_in = {
+		let round_robin = round_robin::RoundRobinBuffer::new(net_in, params.message_buffer_size);
+
+		let round_robin_recovered = round_robin
+			.filter_map(move |(sender, msg)| recovery.check_message(msg).map(move |x| (sender, x)));
+
+		handle_incoming::HandleIncoming::new(
+			params.table.clone(),
+			round_robin_recovered,
+			bft_in_in,
+		).map_err(|_| Error::IoTerminated)
+	};
+
+	let route_messages_out = {
+		let table = params.table.clone();
+		let periodic_table_statements = propagate_statements
+			.or_else(|_| ::futures::future::empty()) // halt the stream instead of error.
+			.map(move |mut batch| { table.fill_batch(&mut batch); batch })
+			.filter(|b| !b.is_empty())
+			.map(OutgoingMessage::Table);
+
+		let complete_out_stream = bft_out_out
+			.map_err(|_| Error::IoTerminated)
+			.map(|bft::ContextCommunication(x)| x)
+			.map(OutgoingMessage::Bft)
+			.select(periodic_table_statements);
+
+		net_out.sink_map_err(|_| Error::IoTerminated).send_all(complete_out_stream)
+	};
+
+	let import_local_candidate = {
+		let table = params.table.clone();
+		local_candidate
+			.into_future()
+			.map(table::Statement::Candidate)
+			.map(Some)
```

Are these two lines moot or am I missing something?

---

it's necessary for the `or_else` in the next line which is currently used to swallow errors in candidate production.